lib/logsource.c

```diff
@@ -44,11 +45,35 @@ log_source_wakeup(LogSource *self)
   msg_diagnostics("Source has been resumed", log_pipe_location_tag(&self->super));
 }
 
+static inline guint32
+_take_reclaimed_window(LogSource *self, guint32 window_size_increment)
+{
+  gssize old = atomic_gssize_sub(&self->window_size_to_be_reclaimed, window_size_increment);
+  gboolean need_to_reclaim = (old > 0);
+
+  if (!need_to_reclaim)
+    {
+      atomic_gssize_sub(&self->window_size_to_be_reclaimed, old);
```

Comment:
What does this mean? I though the inverse of `atomic_gssize_sub(&self->window_size_to_be_reclaimed, window_size_increment)` is needed here, something like:

`atomic_gssize_add(&self->window_size_to_be_reclaimed, window_size_increment);`

Reply:
I think we don't need the inverse... this is about avoiding underflow.
What you suggest is to keep this value on its original value (which is already a _negative_ value, as `old` need to be less than 0). What I do is trying to keep the value 0... But I realized, that this is the wrong place where I'm doing this, so thanks for the question!

There are two competing thread:
a) main thread: timer, sets the `window_size_to_be_reclaimed` when needed _and_ `pending_reclaimed` is 0 (which means that all the requested window slots has been reclaimed)

b) a worker thread (where the ACKs are coming)
In this situation the three (sub/set/add) atomic instructions are 'racy', I mean it is possible that first we decrease the old value (`pending_reclaimed` is 0!) from the worker thread, main thread checks `pending_reclaimed`, it is 0, sets the `window_size_to_be_reclaimed`, then we _increase_ its value... 

The solution is to set it to 0 in the main thread (in the timer callback). When the `pending_reclaimed` is 0 _and_ the `window_size_to_be_reclaimed` is less than 0 _and_ we don't have to reclaim window slots, we have to set the `window_size_to_be_reclaimed` to 0. With this change we can avoid underflow...